graph/src/firehose/endpoints.rs

```diff
@@ -113,49 +168,100 @@ impl FirehoseEndpoint {
     // inside FirehoseEndpoints that is not used (is always cloned).
-    pub fn has_subgraph_capacity(self: &Arc<Self>) -> bool {
+    pub fn get_capacity(self: &Arc<Self>) -> AvailableCapacity {
         self.subgraph_limit
-            .has_capacity(Arc::strong_count(self).saturating_sub(1))
+            .get_capacity(Arc::strong_count(self).saturating_sub(1))
     }
 
-    pub async fn get_block<M>(
+    fn new_client(
         &self,
-        cursor: FirehoseCursor,
-        logger: &Logger,
-    ) -> Result<M, anyhow::Error>
-    where
-        M: prost::Message + BlockchainBlock + Default + 'static,
-    {
-        let token_metadata = match self.token.clone() {
-            Some(token) => Some(MetadataValue::try_from(token.as_str())?),
-            None => None,
+    ) -> FetchClient<
+        InterceptedService<MetricsInterceptor<Channel>, impl tonic::service::Interceptor>,
+    > {
+        let metrics = MetricsInterceptor {
+            metrics: self.endpoint_metrics.cheap_clone(),
+            service: self.channel.cheap_clone(),
+            host: self.host.clone(),
         };
 
-        let mut client = firehose::fetch_client::FetchClient::with_interceptor(
-            self.channel.cheap_clone(),
-            move |mut r: Request<()>| {
-                if let Some(ref t) = token_metadata {
-                    r.metadata_mut().insert("authorization", t.clone());
-                }
+        let mut client: FetchClient<
+            InterceptedService<MetricsInterceptor<Channel>, AuthInterceptor>,
+        > = FetchClient::with_interceptor(metrics, self.auth.clone())
+            .accept_compressed(CompressionEncoding::Gzip);
 
-                Ok(r)
-            },
-        )
-        .accept_compressed(CompressionEncoding::Gzip);
+        if self.compression_enabled {
+            client = client.send_compressed(CompressionEncoding::Gzip);
+        }
+
+        client
+    }
+
+    fn new_stream_client(
+        &self,
+    ) -> StreamClient<
+        InterceptedService<MetricsInterceptor<Channel>, impl tonic::service::Interceptor>,
+    > {
+        let metrics = MetricsInterceptor {
+            metrics: self.endpoint_metrics.cheap_clone(),
+            service: self.channel.cheap_clone(),
+            host: self.host.clone(),
+        };
+
+        let mut client = StreamClient::with_interceptor(metrics, self.auth.clone())
+            .accept_compressed(CompressionEncoding::Gzip);
 
         if self.compression_enabled {
             client = client.send_compressed(CompressionEncoding::Gzip);
         }
 
+        client
+    }
+
+    fn new_substreams_client(
+        &self,
+    ) -> substreams::stream_client::StreamClient<
+        InterceptedService<MetricsInterceptor<Channel>, impl tonic::service::Interceptor>,
+    > {
+        let metrics = MetricsInterceptor {
+            metrics: self.endpoint_metrics.cheap_clone(),
+            service: self.channel.cheap_clone(),
+            host: self.host.clone(),
+        };
```

Comment:
This block of code is repeated a few times, I would suggest a `fn metrics_interceptor(&self) -> MetricsInterceptor<Channel>` method.

Reply:
I think it's going to be almost the exact same code, prolly the best way here would be to write a macro to implement these 3 functions, they are effectively the same with a different generated type